engine/collection/epochmgr/engine.go

```diff
@@ -30,20 +33,56 @@ var ErrUnstakedForEpoch = fmt.Errorf("we are not a staked node in the epoch")
 
 // EpochComponents represents all dependencies for running an epoch.
 type EpochComponents struct {
-	state    cluster.State
-	prop     network.Engine
-	sync     network.Engine
-	hotstuff module.HotStuff
+	*component.ComponentManager
+	state      cluster.State
+	prop       network.Engine
+	sync       network.Engine
+	hotstuff   module.HotStuff
+	aggregator hotstuff.VoteAggregator
 }
 
-// Ready starts all epoch components.
-func (ec *EpochComponents) Ready() <-chan struct{} {
-	return util.AllReady(ec.prop, ec.sync, ec.hotstuff)
+var _ module.Startable = (*EpochComponents)(nil)
+var _ module.ReadyDoneAware = (*EpochComponents)(nil)
+
+func NewEpochComponents(
+	state cluster.State,
+	prop network.Engine,
+	sync network.Engine,
+	hotstuff module.HotStuff,
+	aggregator hotstuff.VoteAggregator,
+) *EpochComponents {
+	components := &EpochComponents{
+		state:      state,
+		prop:       prop,
+		sync:       sync,
+		hotstuff:   hotstuff,
+		aggregator: aggregator,
+	}
+
+	builder := component.NewComponentManagerBuilder()
+	builder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		aggregator.Start(ctx)
+		<-util.AllReady(components.prop, components.sync, components.hotstuff, components.aggregator)
+		ready()
+		<-util.AllDone(components.prop, components.sync, components.hotstuff, components.aggregator)
```

Comment:
Why are we calling `Done` on the epoch components here? `ready()` is non-blocking from [what I can see](https://github.com/onflow/flow-go/blob/8febde2129b92393e96eeca6f9d5d8212cea5f14/module/component/component.go#L241-L244), so we will start (`Ready`) then immediately stop (`Done`) all of these components. I think we should only be calling `Done` on the components when we are tearing them down in `stopEpochComponents`.

Reply:
That is true, a mistake on my side, I have wrongly assumed new semantics for all components, we should wait only for `Startable` components which is `EventLoop` and `VoteAggregator`